File: Data Structures & Algorithms/trapping-rain-water/submission-2.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        lmax, rmax = [0] * n, [0] * n

        for i in range(1, n):
            lmax[i] = max(lmax[i-1], height[i-1])
        
        for i in range(n-2, -1, -1):
            rmax[i] = max(rmax[i+1], height[i+1])
        
        res = 0
        for i in range(n):
            res += max(0, min(lmax[i], rmax[i]) - height[i])
        
        return res

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> l(n, 1), r(n, 1), res(n);

        for(int i=1; i<n; i++){
            l[i] = l[i-1] * nums[i-1];
        }

        for(int j=n-2; j>=0; j--){
            r[j] = r[j+1] * nums[j+1];
        }

        for(int i=0; i<n; i++){
            res[i] = l[i] * r[i];
        }

        return res;
    }
};
